game_template/lib/src/in_app_purchase/in_app_purchase.dart

```diff
@@ -0,0 +1,189 @@
+  ///
+  /// Currently, the only supported in-app purchase is ad removal.
+  /// To support more, ad additional classes similar to [AdRemovalPurchase]
+  /// and modify this method.
+  Future<void> buy() async {
+    if (!await inAppPurchaseInstance.isAvailable()) {
+      _reportError('InAppPurchase.instance not available');
+      return;
+    }
+
+    _adRemoval = const AdRemovalPurchase.pending();
+    notifyListeners();
+
+    _log.info('Querying the store with queryProductDetails()');
+    final response = await inAppPurchaseInstance
+        .queryProductDetails({AdRemovalPurchase.productId});
+
+    if (response.error != null) {
+      _reportError('There was an error when making the purchase: '
+          '${response.error}');
+      return;
+    }
+
+    if (response.productDetails.length != 1) {
+      _log.info(
+        'Products in response: '
+        '${response.productDetails.map((e) => '${e.id}: ${e.title}, ').join()}',
+      );
+      _reportError('There was an error when making the purchase: '
+          'product ${AdRemovalPurchase.productId} does not exist?');
+      return;
+    }
+    final productDetails = response.productDetails.single;
+
+    _log.info('Making the purchase');
+    final purchaseParam = PurchaseParam(productDetails: productDetails);
+    try {
+      final success = await inAppPurchaseInstance.buyNonConsumable(
+          purchaseParam: purchaseParam);
+      _log.info('buyNonConsumable() request was sent with success: $success');
+      // The result of the purchase will be reported in the purchaseStream,
+      // which is handled in [_listenToPurchaseUpdated].
+    } catch (e) {
+      _log.severe(
+          'Problem with calling inAppPurchaseInstance.buyNonConsumable(): '
+          '$e');
+    }
+  }
+
+  @override
+  void dispose() {
+    _subscription?.cancel();
+    super.dispose();
+  }
+
+  /// Asks the underlying platform to list purchases that have been already
+  /// made (for example, in a previous session of the game).
+  Future<void> restorePurchases() async {
+    if (!await inAppPurchaseInstance.isAvailable()) {
+      _reportError('InAppPurchase.instance not available');
+      return;
+    }
+
+    try {
+      await inAppPurchaseInstance.restorePurchases();
+    } catch (e) {
+      _log.severe('Could not restore in-app purchases: $e');
+    }
+    _log.info('In-app purchases restored');
+  }
+
+  /// Subscribes to the [inAppPurchaseInstance.purchaseStream].
+  void subscribe() {
+    _subscription?.cancel();
+    _subscription =
+        inAppPurchaseInstance.purchaseStream.listen((purchaseDetailsList) {
+      _listenToPurchaseUpdated(purchaseDetailsList);
+    }, onDone: () {
+      _subscription?.cancel();
+    }, onError: (dynamic error) {
+      _log.severe('Error occurred on the purchaseStream: $error');
+    });
+  }
+
+  Future<void> _listenToPurchaseUpdated(
+      List<PurchaseDetails> purchaseDetailsList) async {
+    for (var purchaseDetails in purchaseDetailsList) {
```

Comment:
Maybe make this `final` to be consistent with other loops in the sample:

```suggestion
    for (final purchaseDetails in purchaseDetailsList) {
```